File: frontend/src/api/alerts.ts
import { http } from "./http";

// Defensive type so we don’t depend on ../types/api being perfect
export type AlertEvent = {
  id: number;
  device_id?: string | null;
  parameter?: string | null;
  sensor_type?: string | null;
  message?: string | null;

  // common variants
  level?: "INFO" | "WARNING" | "CRITICAL" | string;
  severity?:
    | "BELOW_MIN"
    | "ABOVE_MAX"
    | "info"
    | "warning"
    | "critical"
    | string;
  status?: "active" | "acked" | "resolved" | string;

  value?: number | null;
  min_value?: number | null;
  max_value?: number | null;

  created_at?: string | null;
  acked_at?: string | null;
  ack_note?: string | null;
  resolved_at?: string | null;

  [k: string]: any;
};

export type AlertTrendBucket = {
  ts: string;
  count: number;
  [k: string]: any;
};

export type AlertsQuery = {
  limit?: number;
  offset?: number;

  q?: string;
  device_id?: string;

  // status lifecycle filter
  status?: "active" | "acked" | "resolved" | string;

  // level filter (CRITICAL/WARNING)
  level?: "WARNING" | "CRITICAL" | string;

  // direction or severity filter (BELOW_MIN/ABOVE_MAX) OR older severity strings
  severity?: string;

  // optional convenience boolean
  active_only?: boolean;
};

type AlertsResponse = {
  ok?: boolean;
  count?: number;
  total?: number;

  // most common
  data?: AlertEvent[];

  // alternates
  items?: AlertEvent[];
  alerts?: AlertEvent[];
  message?: string;
  error?: string;

  // sometimes returned as { data: { items: [...] } }
  [k: string]: any;
};

function normalizeAlertsResponse(payloadIn: any) {
  // IMPORTANT:
  // Our http.get() returns the parsed JSON payload directly (not { data: ... } wrapper).
  const payload = payloadIn ?? {};

  const data: AlertEvent[] =
    (Array.isArray(payload?.data) ? payload.data : null) ??
    (Array.isArray(payload?.items) ? payload.items : null) ??
    (Array.isArray(payload?.alerts) ? payload.alerts : null) ??
    (Array.isArray(payload?.data?.items) ? payload.data.items : null) ??
    [];

  const count: number =
    typeof payload?.count === "number"
      ? payload.count
      : typeof payload?.total === "number"
        ? payload.total
        : data.length;

  const ok: boolean =
    typeof payload?.ok === "boolean"
      ? payload.ok
      : // if it looks like a normal list payload, treat as ok
        Array.isArray(data) || typeof payload?.count === "number";

  const error =
    payload?.error ||
    payload?.message ||
    (ok ? undefined : "Request failed (bad payload shape)");

  return { ok, count, data, error: error ? String(error) : undefined };
}

// GET /api/alerts?limit=50&offset=0&status=active&level=CRITICAL&severity=ABOVE_MAX&device_id=...&q=...
export async function fetchAlerts(query: AlertsQuery = {}) {
  const params = new URLSearchParams();

  params.set("limit", String(query.limit ?? 50));
  if (typeof query.offset === "number")
    params.set("offset", String(query.offset));

  const q = query.q?.trim();
  if (q) params.set("q", q);

  const device = query.device_id?.trim();
  if (device) params.set("device_id", device);

  if (query.active_only) params.set("active_only", "1");
  if (query.status) params.set("status", String(query.status));
  if (query.level) params.set("level", String(query.level));
  if (query.severity) params.set("severity", String(query.severity));

  const path = `/api/alerts?${params.toString()}`;

  try {
    const payload = await http.get<AlertsResponse>(path);
    const norm = normalizeAlertsResponse(payload);
    return {
      ok: norm.ok,
      data: norm.data,
      count: norm.count,
      error: norm.error,
      usedPath: path,
    };
  } catch (e: any) {
    return {
      ok: false,
      data: [],
      count: 0,
      error: e?.message || "Failed to load alerts",
      usedPath: path,
    };
  }
}

export async function fetchAlertTrends(
  query: { hours?: number; bucket_min?: number } = {},
) {
  const params = new URLSearchParams();
  if (query.hours != null) params.set("hours", String(query.hours));
  if (query.bucket_min != null)
    params.set("bucket_min", String(query.bucket_min));
  const path = `/api/alerts/trends?${params.toString()}`;

  try {
    const payload = await http.get<{
      ok?: boolean;
      data?: AlertTrendBucket[];
      items?: AlertTrendBucket[];
      error?: string;
      message?: string;
      [k: string]: any;
    }>(path);

    const data = Array.isArray(payload?.data)
      ? payload.data
      : Array.isArray((payload as any)?.items)
        ? (payload as any).items
        : Array.isArray((payload as any)?.data?.items)
          ? (payload as any).data.items
          : [];

    const ok =
      typeof payload?.ok === "boolean"
        ? payload.ok
        : Array.isArray(data) || data.length >= 0;

    return {
      ok,
      data,
      error:
        payload?.error || payload?.message
          ? String(payload.error || payload.message)
          : undefined,
      usedPath: path,
    };
  } catch (e: any) {
    return {
      ok: false,
      data: [],
      error: e?.message || "Failed to load alert trends",
      usedPath: path,
    };
  }
}

// Acknowledge
export async function ackAlert(id: number, note?: string) {
  const body = note ? { ack_note: note } : {};
  try {
    const payload = await http.post<{ ok?: boolean; data?: AlertEvent }>(
      `/api/alerts/${id}/ack`,
      body,
    );
    return ((payload as any)?.data ?? payload) as any;
  } catch {
    const payload = await http.put<{ ok?: boolean; data?: AlertEvent }>(
      `/api/alerts/${id}/ack`,
      body,
    );
    return ((payload as any)?.data ?? payload) as any;
  }
}

// Resolve
export async function resolveAlert(id: number) {
  try {
    const payload = await http.post<{ ok?: boolean; data?: AlertEvent }>(
      `/api/alerts/${id}/resolve`,
      {},
    );
    return ((payload as any)?.data ?? payload) as any;
  } catch {
    const payload = await http.put<{ ok?: boolean; data?: AlertEvent }>(
      `/api/alerts/${id}/resolve`,
      {},
    );
    return ((payload as any)?.data ?? payload) as any;
  }
}

File: app/routes/alerts.py
from flask import Blueprint, request, jsonify

from app.services.alerts_service import (
    list_alerts_page,
    ack_alert,
    resolve_alert,
)
from app.services.alerts_summary_service import get_alert_summary
from app.services.alerts_summary_service import get_alert_trends

alerts_bp = Blueprint("alerts_bp", __name__, url_prefix="/api/alerts")


def _to_int(v):
    if v in (None, ""):
        return None
    return int(v)


# =========================
# GET /api/alerts
# =========================
@alerts_bp.get("")
def get_alerts():
    """
    List alerts (enterprise inbox).
    Filters (all optional):
      ?device_id=ESP32_01
      ?node_id=1
      ?parameter=temperature
      ?status=active|acked|resolved|all
      ?severity=BELOW_MIN|ABOVE_MAX
      ?level=WARNING|CRITICAL
      ?q=free text (message/device/parameter)
      ?limit=50
      ?offset=0

    Legacy flags still supported:
      ?active_only=true
      ?acked_only=true|false
    """
    try:
        device_id = request.args.get("device_id")
        node_id = _to_int(request.args.get("node_id"))
        parameter = request.args.get("parameter")

        # New enterprise filters
        status = (request.args.get("status") or "").strip().lower() or None
        severity = (request.args.get("severity") or "").strip().upper() or None
        level = (request.args.get("level") or "").strip().upper() or None
        q = (request.args.get("q") or "").strip() or None

        # Backward compatible flags
        active_only = (request.args.get("active_only") or "").lower() in {"1", "true", "yes"}

        acked_only_param = request.args.get("acked_only")
        acked_only = None
        if acked_only_param is not None:
            acked_only = acked_only_param.lower() in {"1", "true", "yes"}

        limit = int(request.args.get("limit") or 50)
        offset = int(request.args.get("offset") or 0)

        # If new 'status' is provided, it overrides legacy flags
        if status:
            if status == "active":
                active_only = True
                acked_only = None
            elif status == "acked":
                active_only = True
                acked_only = True
            elif status == "resolved":
                active_only = False
                acked_only = None
            elif status == "all":
                active_only = False
                acked_only = None
            else:
                return jsonify({"ok": False, "error": "Invalid status"}), 400

        items, total = list_alerts_page(
            device_id=device_id,
            node_id=node_id,
            parameter=parameter,
            active_only=active_only,
            acked_only=acked_only,
            severity=severity,
            level=level,
            q=q,
            limit=limit,
            offset=offset,
        )

        return jsonify({"ok": True, "count": int(total), "data": items}), 200

    except Exception:
        return jsonify({"ok": False, "error": "Server error"}), 500


# =========================
# POST /api/alerts/<id>/ack
# =========================
@alerts_bp.post("/<int:alert_id>/ack")
def post_ack_alert(alert_id: int):
    """
    Acknowledge an alert.
    Accepts payload:
      { "note": "..." }  OR  { "ack_note": "..." }
    """
    try:
        payload = request.get_json(silent=True) or {}
        note = payload.get("note", "")
        if not note:
            note = payload.get("ack_note", "")

        data = ack_alert(alert_id, note=note)
        return jsonify({"ok": True, "data": data}), 200
    except ValueError as e:
        return jsonify({"ok": False, "error": str(e)}), 404
    except Exception:
        return jsonify({"ok": False, "error": "Server error"}), 500


# =========================
# POST /api/alerts/<id>/resolve
# =========================
@alerts_bp.post("/<int:alert_id>/resolve")
def post_resolve_alert(alert_id: int):
    """
    Resolve an alert (ACTIVE -> resolved).
    """
    try:
        data = resolve_alert(alert_id)
        return jsonify({"ok": True, "data": data}), 200
    except ValueError as e:
        return jsonify({"ok": False, "error": str(e)}), 404
    except Exception:
        return jsonify({"ok": False, "error": "Server error"}), 500


# =========================
# GET /api/alerts/summary
# =========================
@alerts_bp.get("/summary")
def alerts_summary():
    """
    Dashboard summary endpoint.
    """
    try:
        device_id = request.args.get("device_id")
        node_id = _to_int(request.args.get("node_id"))
        parameter = request.args.get("parameter")

        data = get_alert_summary(device_id=device_id, node_id=node_id, parameter=parameter)
        return jsonify({"ok": True, "data": data}), 200
    except Exception:
        return jsonify({"ok": False, "error": "Server error"}), 500
    

# GET /api/alerts/trends
@alerts_bp.get("/trends")
def alerts_trends():
    try:
        hours = int(request.args.get("hours", 24))
        bucket_min = int(request.args.get("bucket_min", 60))

        device_id = request.args.get("device_id")
        node_id = _to_int(request.args.get("node_id"))
        parameter = request.args.get("parameter")

        data = get_alert_trends(
            hours=hours,
            bucket_min=bucket_min,
            device_id=device_id,
            node_id=node_id,
            parameter=parameter,
        )
        return jsonify({"ok": True, "data": data}), 200
    except Exception:
        return jsonify({"ok": False, "error": "Server error"}), 500

    